hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java

```diff
@@ -366,40 +366,62 @@ public void connectionAdded(Connection conn) {
         @Override
         public void connectionRemoved(Connection connection) {
             if (connection.isClient() && nodeEngine.isRunning()) {
-                ClientEndpointImpl endpoint = (ClientEndpointImpl) endpointManager.getEndpoint(connection);
+                final ClientEndpointImpl endpoint = (ClientEndpointImpl) endpointManager.getEndpoint(connection);
                 if (endpoint == null) {
+                    logger.finest("connectionRemoved: No endpoint for connection:" + connection);
                     return;
                 }
 
                 if (!endpoint.isFirstConnection()) {
+                    logger.finest("connectionRemoved: Not the owner conn:" + connection + " for endpoint " + endpoint);
                     return;
                 }
 
                 String localMemberUuid = node.getLocalMember().getUuid();
                 String ownerUuid = endpoint.getPrincipal().getOwnerUuid();
                 if (localMemberUuid.equals(ownerUuid)) {
-                    callDisconnectionOperation(endpoint);
+                    try {
+                        nodeEngine.getExecutionService().schedule(new Runnable() {
+                            @Override
+                            public void run() {
+                                callDisconnectionOperation(endpoint);
+                            }
+                        }, ENDPOINT_REMOVE_DELAY_SECONDS, TimeUnit.SECONDS);
+                    } catch (RejectedExecutionException e) {
+                        if (logger.isFinestEnabled()) {
+                            logger.finest(e);
+                        }
+                    }
                 }
             }
         }
 
         private void callDisconnectionOperation(ClientEndpointImpl endpoint) {
             Collection<Member> memberList = nodeEngine.getClusterService().getMembers();
             OperationService operationService = nodeEngine.getOperationService();
-            ClientDisconnectionOperation op = createClientDisconnectionOperation(endpoint.getUuid());
+            String memberUuid = getLocalMember().getUuid();
+            String clientUuid = endpoint.getUuid();
+
+            String ownerMember = ownershipMappings.get(clientUuid);
+            if (memberUuid != ownerMember) {
```

Comment:
Member is uniquely identified by the uuid, hence i thought that it is OK to compare uuid strings.